android/src/main/new_arch/react/renderer/components/RNLiveMarkdownSpec/MarkdownTextInputDecoratorShadowNode.cpp

```diff
@@ -0,0 +1,125 @@
+
+  const auto markdownStyle =
+      this->getProps()->rawProps["markdownStyle"];
+  const auto parserId =
+      this->getProps()->rawProps["parserId"].asInt();
+
+  const auto decoratorPropsRNM =
+      ReadableNativeMap::newObjectCxxArgs(markdownStyle);
+  const auto decoratorPropsRM =
+      jni::make_local(reinterpret_cast<ReadableMap::javaobject>(
+                          decoratorPropsRNM.get()));
+
+  const auto &fabricUIManager =
+      this->getContextContainer()->at<JFabricUIManager::javaobject>("FabricUIManager");
+
+  const auto customUIManager = jni::make_global(createMethod(
+      customFabricUIManagerClass, fabricUIManager,
+      decoratorPropsRM.get(), parserId));
+  const auto contextContainer =
+      std::make_shared<ContextContainer const>();
+  contextContainer->insert("FabricUIManager", customUIManager);
+
+  customContextContainer_ = contextContainer;
+  previousMarkdownStyle_ = markdownStyle;
+  previousParserId_ = parserId;
+}
+
+void MarkdownTextInputDecoratorShadowNode::updateCustomContextContainerIfNeeded() {
+  const auto markdownStyle =
+      this->getProps()->rawProps["markdownStyle"];
+  const auto parserId =
+      this->getProps()->rawProps["parserId"].asInt();
+
+  if (parserId != previousParserId_ || markdownStyle != previousMarkdownStyle_) {
+    createCustomContextContainer();
+  }
+}
+
+void MarkdownTextInputDecoratorShadowNode::overwriteTextLayoutManager() {
+  const auto &children = getChildren();
+  if (children.empty()) {
+    return;
+  }
+  react_native_assert(
+      children.size() == 1 &&
+      "MarkdownTextInputDecoratorView received more than one child");
+
+  const auto child = std::dynamic_pointer_cast<const AndroidTextInputShadowNode>(
+      children.at(0));
+
+  react_native_assert(
+      child != nullptr &&
+      "MarkdownTextInputDecoratorView received child other than a TextInput");
+  child->ensureUnsealed();
+
+  const auto mutableChild = std::const_pointer_cast<AndroidTextInputShadowNode>(child);
+  mutableChild->setTextLayoutManager(
+      std::make_shared<TextLayoutManager>(customContextContainer_));
+}
+
+void MarkdownTextInputDecoratorShadowNode::appendChild(
+    const ShadowNode::Shared &child) {
+  YogaLayoutableShadowNode::appendChild(child);
+
+  overwriteTextLayoutManager();
+}
+
+void MarkdownTextInputDecoratorShadowNode::replaceChild(
+    const ShadowNode &oldChild, const ShadowNode::Shared &newChild,
+    size_t suggestedIndex) {
+  YogaLayoutableShadowNode::replaceChild(oldChild, newChild, suggestedIndex);
+
+  overwriteTextLayoutManager();
+}
+
+void MarkdownTextInputDecoratorShadowNode::layout(LayoutContext layoutContext) {
+  YogaLayoutableShadowNode::layout(layoutContext);
+
+  const auto &children = getChildren();
+  react_native_assert(
+      children.size() == 1 &&
+      "MarkdownTextInputDecoratorView didn't receive exactly one child");
+
+  auto child = std::static_pointer_cast<const YogaLayoutableShadowNode>(getChildren()[0]);
```

Comment:
Once we confirm that `dynamic_pointer_cast` in `overwriteTextLayoutManager` indeed returns a non-null `child` we can use `static_pointer_cast` here, is that correct? Why don't we just use `dynamic_cast` everywhere?